### PR TITLE
ansible CLI allow other types than a string in module arguments 

### PR DESCRIPTION
**Pull request by mnecas**
##### SUMMARY

Hello there,
when the user is trying to add module arguments with only using the CLI he can only pass string values.
With this patch, if the user would use only the JSON it would format the objects properly, otherways it will stay the same. 

Examples:
This stays the same
```cmd
ansible -m debug localhost -a 'msg={"msg":[{"test":123}]}'
```
```json
"msg": "{\"msg\":[{\"test\":123}]}"
```
This example always throw error before the change
```cmd
ansible -m debug localhost -a '{"msg":[{"test":123}]}'
```
```json
    "msg": [
        {
            "test": 123
        }
    ]
```

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ansible CLI 
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by mnecas** [comment]
@sivel @bcoca could you please review?

File: lib/ansible/cli/adhoc.py [hunk 1]
```diff
@@ -64,10 +64,18 @@ def post_process_args(self, options):
 
         return options
 
-    def _play_ds(self, pattern, async_val, poll):
-        check_raw = context.CLIARGS['module_name'] in C.MODULE_REQUIRE_ARGS
-
-        mytask = {'action': {'module': context.CLIARGS['module_name'], 'args': parse_kv(context.CLIARGS['module_args'], check_raw=check_raw)},
+    def _play_ds(self, pattern, async_val, poll, loader):
+        if context.CLIARGS['module_args'].startswith(u"@"):
```
Comment (bcoca) [inline lib/ansible/cli/adhoc.py]: at this point, just use a playbook

File: lib/ansible/cli/adhoc.py [hunk 2]
```diff
@@ -64,10 +64,18 @@ def post_process_args(self, options):
...
+            # Argument is a YAML file (JSON is a subset of YAML)
+            args = loader.load_from_file(context.CLIARGS['module_args'][1:])
+        elif context.CLIARGS['module_args'][0] in [u'/', u'.']:
+            raise AnsibleOptionsError("Please prepend extra_vars filename '%s' with '@'" % context.CLIARGS['module_args'])
+        elif context.CLIARGS['module_args'][0] in [u'[', u'{']:
+            args = loader.load(context.CLIARGS['module_args'])
+        else:
+            check_raw = context.CLIARGS['module_name'] in C.MODULE_REQUIRE_ARGS
+            args = parse_kv(context.CLIARGS['module_args'], check_raw=check_raw)
```
Comment (bcoca) [inline lib/ansible/cli/adhoc.py:74]: I would reverse order and have parse_kv go first, since it will also still be the most common usage, also the current detection will create many issues, we have used it before (still do for safe_eval).

Use  a try/fail approach on parse_kv to try yaml parsing afterwards and error out if all fail

the module_require_args check should not be limted to kv, but also force skip yaml attempt

File: lib/ansible/cli/adhoc.py [hunk 3]
```diff
@@ -64,10 +64,18 @@ def post_process_args(self, options):
 
         return options
 
-    def _play_ds(self, pattern, async_val, poll):
-        check_raw = context.CLIARGS['module_name'] in C.MODULE_REQUIRE_ARGS
-
-        mytask = {'action': {'module': context.CLIARGS['module_name'], 'args': parse_kv(context.CLIARGS['module_args'], check_raw=check_raw)},
+    def _play_ds(self, pattern, async_val, poll, loader):
```
Comment (bcoca) [inline lib/ansible/cli/adhoc.py]: do not use loader, there are from_yaml/json functions you should use instead, loader has many other implications i would not want to apply to this case.

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/cli/adhoc.py (hunk 3 above)
Comment (mnecas) [reply lib/ansible/cli/adhoc.py]: ack will use basic json.load

File: lib/ansible/cli/adhoc.py (hunk 2 above)
Comment (mnecas) [reply lib/ansible/cli/adhoc.py:74]: the try/fail method will not work because the parse_kv at that moment does not fail but it creates raw params (`{'_raw_params': '{"msg":[{"test":123}]}'}`)

File: lib/ansible/cli/adhoc.py (hunk 1 above)
Comment (mnecas) [reply lib/ansible/cli/adhoc.py]: I agree with removal, I added it so it matches the extra vars, but I'm a bit afraid there will be some more edge cases

**Comment by bcoca** [comment]
also you can always:

``` ansible -m debug localhost -a 'msg={{ [{"test":123}]|list}}'```

**Comment by mnecas** [comment]
> also you can always:
> 
> ` ansible -m debug localhost -a 'msg={{ [{"test":123}]|list}}'`

ohhh use the `{{ }}` around the dict I missed that usage, thanks a lot!
This works so I don't mind closing the PR.
